### PR TITLE
Make downvotes actually reduce user karma.

### PR DESCRIPTION
**Pull request by zorbathut**

### CODE REVIEW

**Comment by zorbathut** [comment]
Fixes #116, maybe?

**Comment by zorbathut** [comment]
Alright, note relative to #116:

"Currently, true votes are tracked in the voting tables, but for the purposes of coins/karma and thread/comment score and sorting, downvotes are counted as upvotes."

As near as I can tell, this isn't true for *everything*, but only for total user karma. I might've missed something though, so, uh, another lookover if you would?

**Comment by JulianRota** [comment]
It's a start, but this also needs to fix the [un-voting code here](https://github.com/themotte/rDrama/blob/17d550f00e98eb023cbd0d1d44844b67532740bc/files/routes/votes.py#L84) and the [calculation of `realupvotes` here](https://github.com/themotte/rDrama/blob/17d550f00e98eb023cbd0d1d44844b67532740bc/files/routes/votes.py#L113), which is what is actually used for sorting posts and comments.

I had been trying to figure out how to do the correct query for `realupvotes` in SQLAlchemy - I think it ought to be possible to do a sum in the database.

**Comment by zorbathut** [comment]
Ah hey, you're right! I will fix those, uh, tomorrow. Good catch though :D

File: files/routes/votes.py [hunk 1]
```diff
@@ -56,62 +56,72 @@ def admin_vote_info_get(v):
...
+	g.db.add(post.author)
+	g.db.add(vote)
+	
+	# update post stats (this is horrendously slow?!)
 	g.db.flush()
 	post.upvotes = g.db.query(Vote.submission_id).filter_by(submission_id=post.id, vote_type=1).count()
 	post.downvotes = g.db.query(Vote.submission_id).filter_by(submission_id=post.id, vote_type=-1).count()
-	post.realupvotes = g.db.query(Vote.submission_id).filter_by(submission_id=post.id, real=True).count()
-	if post.author.progressivestack: post.realupvotes *= 2
+	post.realupvotes = g.db.query(Vote.submission_id).filter_by(submission_id=post.id, vote_type=1, real=True).count()
```
Comment (JulianRota) [inline files/routes/votes.py:122]: Based on the way we use `realupvotes` to sort [here](https://github.com/themotte/rDrama/blob/frost/files/routes/front.py#L347) (and similarly for comments in routes/posts.py), I think this calculation needs to take downvotes into account too.

File: files/routes/votes.py [hunk 2]
```diff
@@ -56,62 +56,72 @@ def admin_vote_info_get(v):
...
 
+	if vote:
+		# remove the old score data
+		if points_matter:
+			post.author.coins -= vote.vote_type
+			post.author.truecoins -= vote.vote_type
+			# we'll be saving later anyway, so don't bother doing so here
+	else:
+		# create new vote data
 		DEFAULT_IMAGE = '/assets/images/default-profile-pic.webp'
```
Comment (JulianRota) [inline files/routes/votes.py]: Not really sure what they were doing with this but I guess we don't need it anymore

File: files/routes/votes.py [hunk 3]
```diff
@@ -56,62 +56,72 @@ def admin_vote_info_get(v):
...
-		else: real = True
-
-		vote = Vote(user_id=v.id,
-					vote_type=new,
-					submission_id=post_id,
-					app_id=v.client.application.id if v.client else None,
-					real = real
+		vote = Vote(
+						user_id=v.id,
+						vote_type=new,
```
Comment (JulianRota) [inline files/routes/votes.py:96]: I suppose there's no need to set `vote_type` here when we're also setting it below

Comment (zorbathut) [reply files/routes/votes.py:96]: Yeah, exactly; there was actually a bug in the original one that the `real` flag would never be changed after the first setting.

File: files/routes/votes.py (hunk 1 above)
Comment (zorbathut) [reply files/routes/votes.py:122]: So I actually think that `realupvotes` was never meant to include downvotes; it actually is "real upvotes", the name is accurate. Note that "top", "bottom", "controversial", and so forth all take downvotes into account separately, and I believe a similar thing is happening here. I haven't sat down to figure out whatever the hell is going on with that "top" equation, but I think it is literally *just* meant to consider upvotes, not downvotes.

(There's a good chance these are all getting changed at some point :V)

File: files/routes/votes.py (hunk 2 above)
Comment (zorbathut) [reply files/routes/votes.py]: I think it was just a really ugly logic ball intended to only change the values once.

**Review by JulianRota — APPROVED** [review]

**Pull request merged**